### PR TITLE
Add TLS min version flag

### PR DESCRIPTION
**Pull request by deads2k**
Adds a flag for controlling the minimum TLS level allowed.

/assign liggitt

@kubernetes/sig-node-pr-reviews @k8s-mirror-api-machinery-pr-reviews 

```release-note
--tls-min-version on kubelet and kube-apiserver allow for configuring minimum TLS versions
```

### CODE REVIEW

File: cmd/kubelet/app/options/options.go [hunk 1]
```diff
@@ -446,6 +446,10 @@ func AddKubeletConfigFlags(fs *pflag.FlagSet, c *kubeletconfig.KubeletConfigurat
 		"Comma-separated list of cipher suites for the server. "+
 			"Values are from tls package constants (https://golang.org/pkg/crypto/tls/#pkg-constants). "+
 			"If omitted, the default Go cipher suites will be used")
+	fs.StringVar(&c.TLSMinVersion, "tls-min-version", c.TLSMinVersion,
+		"Minimum TLS version supported. "+
+			"Value must match version names from https://golang.org/pkg/crypto/tls/#pkg-constants. "+
+			"If omitted, the default Go cipher suites will be used")
```
Comment (liggitt) [inline cmd/kubelet/app/options/options.go]: don't promise a default... we want the freedom to change this as needed

File: cmd/kubelet/app/server.go [hunk 2]
```diff
@@ -627,12 +627,20 @@ func InitializeTLS(kf *options.KubeletFlags, kc *kubeletconfiginternal.KubeletCo
 		return nil, err
 	}
 
+	// Can't use SSLv3 because of POODLE and BEAST
+	// Can't use TLSv1.0 because of POODLE and BEAST using CBC cipher
+	// Can't use TLSv1.1 because of RC4 cipher usage
+	minTLSVersion := uint16(tls.VersionTLS12)
+	if len(kc.TLSMinVersion) > 0 {
+		minTLSVersion, err = flag.TLSCipherSuite(kc.TLSMinVersion)
```
Comment (liggitt) [inline cmd/kubelet/app/server.go]: this isn't a cipher suite, it's a TLS version... need a new TLSVersion func

File: staging/src/k8s.io/apiserver/pkg/util/flag/ciphersuites_flag.go [hunk 3]
```diff
@@ -54,11 +54,19 @@ func TLSCipherSuites(cipherNames []string) ([]uint16, error) {
...
+		intValue, err := TLSCipherSuite(cipher)
+		if err != nil {
+			return nil, err
 		}
 		ciphersIntSlice = append(ciphersIntSlice, intValue)
 	}
 	return ciphersIntSlice, nil
 }
+
+func TLSCipherSuite(cipherName string) (uint16, error) {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/util/flag/ciphersuites_flag.go]: this should be `TLSVersion(tlsVersion string) (uint16, error)` that deals with these constants:

```
        VersionSSL30 = 0x0300
        VersionTLS10 = 0x0301
        VersionTLS11 = 0x0302
        VersionTLS12 = 0x0303
```

**Comment by deads2k** [comment]
Ah, the second magic block in openshift crypto.

**Comment by deads2k** [comment]
fixed

File: cmd/kubelet/app/server.go [hunk 4]
```diff
@@ -627,12 +627,14 @@ func InitializeTLS(kf *options.KubeletFlags, kc *kubeletconfiginternal.KubeletCo
...
 	}
 
+	minTLSVersion, err := flag.TLSVersion(kc.TLSMinVersion)
+	if err != nil {
+		return nil, err
+	}
+
 	tlsOptions := &server.TLSOptions{
 		Config: &tls.Config{
-			// Can't use SSLv3 because of POODLE and BEAST
```
Comment (liggitt) [inline cmd/kubelet/app/server.go]: move comments up with the default to explain the choice

Comment (liggitt) [reply cmd/kubelet/app/server.go]: actually, use `DefaultTLSVersion()`?

Comment (liggitt) [reply cmd/kubelet/app/server.go]: actually, that's done internally, nm

File: staging/src/k8s.io/apiserver/pkg/util/flag/ciphersuites_flag.go [hunk 5]
```diff
@@ -62,3 +62,26 @@ func TLSCipherSuites(cipherNames []string) ([]uint16, error) {
 	}
 	return ciphersIntSlice, nil
 }
+
+var versions = map[string]uint16{
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/util/flag/ciphersuites_flag.go]: update TestConstantMaps to keep this in sync

**Comment by liggitt** [comment]
nit on test, LGTM otherwise. can clean up flag help (listing valid values) in followup along with the valid values for ciphers

**Comment by liggitt** [comment]
and needs a release note

**Comment by deads2k** [comment]
test and release note added.

**Comment by derekwaynecarr** [comment]
kubelet changes LGTM

/approve no-issue

**Pull request merged**